### PR TITLE
LOOP-1144 Refactor dosing decision errors to allow serialization

### PR DESCRIPTION
**Pull request by darinkrauss**
- https://tidepool.atlassian.net/browse/LOOP-1144
- Move StoredDosingDecision Codable from LoopKit to Loop to support error serialization
- Extend DosingDecisionStore to store and retrieve StoredDosingDecision
- Use static encoder and decoder for StoredDosingDecision
- Implement Codable for CarbStoreError, DoseStoreError, PumpManagerError, SetBolusError, and LoopError
- Refactor Codable for StoredDosingDecision to support error serialization
- Enable MockService in non-debug builds
- Capture additional dosing decisions with errors

@ps2, @nhamming I would like two reviews on this PR.

### CODE REVIEW

File: LoopTests/Extensions/CarbStoreTests.swift [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    func testCodableFetchError() throws {
+        try assertCarbStoreErrorCodable(.noData)
+    }
+
+    func assertCarbStoreErrorCodable(_ original: CarbStore.CarbStoreError) throws {
+        let data = try PropertyListEncoder().encode(TestContainer(carbStoreError: original))
+        let decoded = try PropertyListDecoder().decode(TestContainer.self, from: data)
+        XCTAssertEqual(decoded.carbStoreError, original)
```
Comment (rickpasetto) [inline LoopTests/Extensions/CarbStoreTests.swift:66]: [nit] I've seen this style of unit test before.  It's a fine test to include, but won't catch problems where the same issue exists in both the encoder & decoder (so common due to the copy/paste nature of writing Codable implementations).  So, please consider this as just a "thought exercise" comment...on strategies I've done in the past, that is almost just as easy to write:

I'll write the encoder test comparing the encoded value against a hardcoded string. Then, I'll write the decoder test that decodes that string and test for object equality (to the original object I encoded in the first test).  I'd "cheat" by letting the encoder generate the string, I'll paste it into the test file, and visually inspect it for correctness.  That way, at least, there's one level of "intermediary" to catch certain coder bugs.  

To be clear, I'm not asking you to change this PR, but rather putting an idea in here for your consideration.

WDYT?

**Review by rickpasetto — APPROVED** [review]
LGTM.  I put a comment in here about unit testing that I'm interested in your thoughts on, but please do not consider it blocking.

File: LoopTests/Extensions/CarbStoreTests.swift (hunk 1 above)
Comment (darinkrauss) [reply LoopTests/Extensions/CarbStoreTests.swift:66]: Yea, I see your point. I think that makes sense when the encoded output is used for any other reason that decoding into an equivalent object later. For example, the encoded JSON output is sent to a JSON API.

In this case it really doesn't matter what the intermediate form is, whether there is a typo, or even if it doesn't look like I thought it would. If there is an object and it is encoded and decoded into another object, then as long as the two objects are "equal" (however that is defined for the use case) we're good.

Comment (rickpasetto) [reply LoopTests/Extensions/CarbStoreTests.swift:66]: Just a counterargument: If there was a bug in both the encoder and decoder returned completely empty objects, they also would be equal.  (This is an `enum`, so that is not really possible, but for a `struct` with mostly optional values, it could happen...and has happened to me in the past).

Comment (darinkrauss) [reply LoopTests/Extensions/CarbStoreTests.swift:66]: True. But I'm only testing the net result of encoding and decoding to see if it gets back to the same as the original object. 

Also, the original object created for each test explicitly do not use `nil` for any property in the struct (or associated value in an enum) because a `nil` in the final object could either be correct or the result of a missing/incorrect encoding/decoding.

In the end, each test builds a fully populated original struct/enum without any `nil`s. The test then encodes that original object into Data and then decodes that Data into a result object. Finally, the result object is compared against the original object to make sure they are the equivalent.

I suppose it is possible that the original object is modified during the encoding process, but, at least for the data structures I'm testing, they are structs or enums and the  `encode` function is not marked as `mutable`.

**Review by nhamming — APPROVED** [review]
LGTM

File: LoopTests/Extensions/CarbStoreTests.swift (hunk 1 above)
Comment (ps2) [reply LoopTests/Extensions/CarbStoreTests.swift:66]: Another argument for putting the serialized string into the test is to make us more aware of changes that might need to be handled via migration. At some point, if multiple versions of the serialized form could exist on disk, we'll want to test old and new versions, and probably won't have the encoder for the old form.

**Review by ps2 — APPROVED** [review]
LGTM!

File: LoopTests/Extensions/CarbStoreTests.swift (hunk 1 above)
Comment (darinkrauss) [reply LoopTests/Extensions/CarbStoreTests.swift:66]: Honestly, my hesitancy in doing this work is that there are, at last count, 64 tests I'd have to update.

Comment (rickpasetto) [reply LoopTests/Extensions/CarbStoreTests.swift:66]: @darinkrauss Like I said originally, this is still a fine test to include.  I'm only making a suggested improvement.  Also, hopefully I was clear, that this was not a blocking request.  This test (and the other tests) can be added to over time, IMO.

Comment (darinkrauss) [reply LoopTests/Extensions/CarbStoreTests.swift:66]: Updated all recent coding tests to check intermediate serialized data string.

**Comment by darinkrauss** [comment]
@rickpasetto Would you mind reviewing the latest coding test changes and re-approving?

**Review by rickpasetto — APPROVED** [review]
LGTM 🔎
